pipeline/train/train.sh

```diff
@@ -30,42 +35,90 @@ mkdir -p "${model_dir}/tmp"
+for index in "${!datasets[@]}"; do
+    train_set_prefix="${datasets[index]}"
+    # OpusTrainer supports only tsv and gzip
+    # TODO: pigz is not found
+    tsv_dataset="${train_set_prefix}.${src}${trg}.tsv" #.gz"
+
+    if [ "${alignments}" != "None" ] ; then
+      train_aln="${alns[index]}"
+      # alignments are currently used only for student that uses one stage of training
+      echo "### Generating tsv dataset with alignments ${alignments}"
+      paste <(${COMPRESSION_CMD} -dc "${train_set_prefix}.${src}.${ARTIFACT_EXT}") \
+            <(${COMPRESSION_CMD} -dc "${train_set_prefix}.${trg}.${ARTIFACT_EXT}") \
+            <(${COMPRESSION_CMD} -dc "${train_aln}") \
+            >"${tsv_dataset}"
+#            | pigz -c >"${tsv_dataset}"
+      rm "${train_aln}"
+      # when using tsv, marian requires --guided-alignments argument to be an index of the alignments in the tsv file
+      extra_params+=("--guided-alignment" "2")
+    else
+      echo "### Generating tsv dataset"
+      # OpusTrainer supports only tsv and gzip
+      paste <(${COMPRESSION_CMD} -dc "${train_set_prefix}.${src}.${ARTIFACT_EXT}") \
+            <(${COMPRESSION_CMD} -dc "${train_set_prefix}.${trg}.${ARTIFACT_EXT}") \
+            >"${tsv_dataset}"
+#            | pigz -c >"${tsv_dataset}"
+    fi
+    # free disk space
+    rm "${train_set_prefix}.${src}.${ARTIFACT_EXT}"
+    rm "${train_set_prefix}.${trg}.${ARTIFACT_EXT}"
+    # replace the dataset path in the template in place
+    sed -i -e "s#<dataset${index}>#${tsv_dataset}#g" "${new_config}"
+done
 
-# if doesn't fit in RAM, remove --shuffle-in-ram and add --shuffle batches
-
-# Marian doesn't support zst natively; we need to decompress before passing them
-# along.
-if [ "${ARTIFACT_EXT}" = "zst" ]; then
-  zstdmt --rm -d "${train_set_prefix}.${src}.${ARTIFACT_EXT}"
-  zstdmt --rm -d "${train_set_prefix}.${trg}.${ARTIFACT_EXT}"
-  zstdmt --rm -d "${valid_set_prefix}.${src}.${ARTIFACT_EXT}"
-  zstdmt --rm -d "${valid_set_prefix}.${trg}.${ARTIFACT_EXT}"
-  ARTIFACT_EXT=""
-else
-  ARTIFACT_EXT=".gz"
-fi
-
-"${MARIAN}/marian" \
-  --model "${model_dir}/model.npz" \
-  -c "configs/model/${model_type}.yml" "configs/training/${model_type}.${training_type}.yml" \
-  --train-sets "${train_set_prefix}".{"${src}","${trg}"}${ARTIFACT_EXT} \
-  -T "${model_dir}/tmp" \
-  --shuffle-in-ram \
-  --vocabs "${vocab}" "${vocab}" \
-  -w "${WORKSPACE}" \
-  --devices ${GPUS} \
-  --sharding local \
-  --sync-sgd \
-  --valid-metrics "${best_model_metric}" ${all_model_metrics[@]/$best_model_metric} \
-  --valid-sets "${valid_set_prefix}".{"${src}","${trg}"}${ARTIFACT_EXT} \
-  --valid-translation-output "${model_dir}/devset.out" \
-  --quiet-translation \
-  --overwrite \
-  --keep-best \
-  --log "${model_dir}/train.log" \
-  --valid-log "${model_dir}/valid.log" \
-  "${extra_params[@]}"
+# Replace the path to vocab (required for alignments to work)
+sed -i -e "s#<vocab>#${vocab}#g" "${new_config}"
```

Comment:
TIL -i behaves differently in MacOS...

Reply:
I ran this on MacOS and it worked. But I don't know if we even want this to be testable locally. It's just too hard to set up the environment, get all the input files etc.

Reply:
Thought: Breaking the problem down into smaller pieces will make parts of it testable so we have confidence on this type of code. Right now my confidence for correctness is pretty low, but I think the risk for merging here is pretty low as well.

I've run into plenty of issues myself where macOS's sed behavior is different from Linux.